[clangd] Add option to fuzzy-match macros in code-complete

MythreyaK opened the pull request:
Adds `MacroFilterPolicy` 

```yaml
Completion:
  MacroFilter: ExactPrefix (default), FuzzyMatch
```

Fixes clangd/clangd#1480

MythreyaK committed:
[clangd] Fuzzy-match some macros

Allow results from macros that do not have leading
or trailing underscores.

MythreyaK committed:
Add config options + tests

MythreyaK committed:
Tweak test

MythreyaK committed:
code-review: update release notes

HighCommander4 commented on clang-tools-extra/clangd/CodeComplete.cpp:
```diff
@@ -2037,14 +2038,34 @@ class CodeCompleteFlow {
+      else
+        return std::nullopt;
+    case MacroFilterPolicy::FuzzyMatch:
+      if (!C.Name.starts_with_insensitive("_") &&
```
If the pattern starts with an underscore (i.e. the user is doing `_foo^`), we should still allow exact prefix matches.

(Some people work on projects like stdlib implementations, I'd bet they would notice if this stopped working ;))

Worth a test case as well.

MythreyaK replied:
Do I add that as a separate option, since most wouldn't be working on stdlib implementations? 

> Worth a test case as well.

Will do :+1:

HighCommander4 reviewed (changes requested):
Thanks for the patch!

Looks good with a minor tweak to the behaviour.

MythreyaK committed:
Simplify filtering logic

If fuzzy-matching, generally exclude macros with
trailing and leading underscores, unless the input
itself begins with an underscore

Co-authored-by: HighCommander4 <zeratul976@hotmail.com>

MythreyaK committed:
add newline to rst

HighCommander4 reviewed (approved):
Thanks, LGTM!
